[FancyZones Editor] Add Fluent Design and theming support

niels9001 opened the pull request:
## Summary of the Pull Request

This PR introduces a brand new UX for the FZ editor.

New features:
- Ability to duplicate layouts
- New UX that simplifies the creation and selection of (new) layouts.
- Theming support (dark/light and all high contrast themes).
- UI can now be fully controlled with tab / arrow keys
- Monitor selection now supports keyboard (for selection)
- Accessibility improvements.
- Other minor enhancements (e.g. scrollable list).


### **Things still to be improved / known issues (before merging):**
* [ ] Flyout menu should exclude certain properties based on layout type (e.g. templates cannot be deleted)
* [ ] Layout previews do not always update after editing. Needs a 'hard refresh' of the dialog.
* [ ] Settings in layout flyout do not work.
* [ ] Duplicate custom layouts does not work (yet).
* [ ] On mouse leave, the preview should reset to the applied layout.
* [ ] Flyout should appear when applied and right-clicking or using Shift+F10


![FZeditor_dark](https://user-images.githubusercontent.com/9866362/101410242-5b90a280-38df-11eb-834a-8365453b8429.gif)


Editor
![image](https://user-images.githubusercontent.com/9866362/100554204-6457f780-3293-11eb-9cc0-9ed5d35d8e66.png)

New layout dialog (heavily inspired by a @zeealeid concept!!!)
![image](https://user-images.githubusercontent.com/9866362/101407850-cb9d2980-38db-11eb-82b8-3ec3e89c7751.png)

**Layout edit flyout**
![image](https://user-images.githubusercontent.com/9866362/101408324-8b8a7680-38dc-11eb-83cd-01af79b37710.png)

**Grid layout editor:**
![image](https://user-images.githubusercontent.com/9866362/101407928-e5d70780-38db-11eb-8c92-6dee68fc3af2.png)

**Canvas layout editor:**
![image](https://user-images.githubusercontent.com/9866362/101408051-16b73c80-38dc-11eb-9a84-de6c09e29c01.png)

**Light theme**
![image](https://user-images.githubusercontent.com/9866362/101408437-af4dbc80-38dc-11eb-95f7-3d7ba80d8890.png)



## PR Checklist
* [X] Applies to #891 #8160 #7075 #7073 #7071 #7048 #6780 #6098 #3672 #3095 #2802 #2588 #2208 #738 #326 #8315 #6867 #8564, #8148 
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

yuyoyuppe commented:
rebased on the current master

enricogior commented:
@niels9001 
we had to rebase your current PR to unblock CI.
Before making more changes, you will need to pull from remote to update your local branch.

enricogior commented:
@niels9001 
you need to remove `MahApps.Metro.dll` from `installer\PowerToysSetup\Product.wxs` line 375:

```
installer\PowerToysSetup\Product.wxs(375,0): Error LGHT0103: The system cannot find the file 'D:\a\1\s\installer\PowerToysSetup\..\..\x64\Release\modules\FancyZones\MahApps.Metro.dll'.
```

htcfreek commented:
I don't know how the editor is looking now.

But maybe it's interesting for you to know the following: After the changes made with 0.27 in the "old" editor, the edit layout button is to small for the German translation.

htcfreek commented:
Feedback:
1. The color sheme of the "New Layout Dialogue" doesn't fit to the other dialoges.
2. We should show the text on the reset button too.
3. I don't like the additional "settings" menu. But unfortunately at the moment I have no idea how to make it better. Maybe a centered dialog overlay could help.
4. We should add a menu indicator like the three dots on the layouts.

niels9001 commented:
![test](https://user-images.githubusercontent.com/9866362/101213517-cb522380-367a-11eb-9d6a-9c43be85e3a0.png)

htcfreek commented:
The spaces are not saved per layout. They are saved per monitor. So we should split the settings for the layout and the space.

enricogior commented:
@htcfreek 
> The spaces are not saved per layout.

The current implementation is broken, we are going to fix it allowing to edit the spaces per layout.

htcfreek commented:
> @htcfreek 
> > The spaces are not saved per layout.
> 
> The current implementation is broken, we are going to fix it allowing to edit the spaces per layout.

Okay. Then we have an other situation and @miels9001's layout makes sense.

alannt777 reviewed (commented):
@niels9001

The new Editor is too large when I tested it on my small laptop. The bottom part of the Editor was not visible

niels9001 commented:
@alannt777 Can you take a screenshot? We might want to make the dialog (somewhat) resizeable. Still working on that.

alannt777 commented:
See this:

![FZ Editor screenshot](https://user-images.githubusercontent.com/67918878/101236457-583ac280-36f7-11eb-95cf-bc269f515875.png)

alannt777 commented:
Also I can't seem to edit my layouts. I guess you might have not implemented that part

alannt777 commented:
@niels9001 I have noticed a bug where if you have only one monitor, you can't edit the layouts

alannt777 commented:
You can't right-click on any layouts at all

niels9001 commented:
@alannt777 please note that this is still in draft and work in progress. There are still bugs I'm actively working on.

alannt777 commented:
I know 😄 

I was just letting you know

Jay-o-Way commented:
@htcfreek had a few very good points, I actually agree with all of them!
About the reset icon: you actually did use the "refresh" icon. `E7A7` "undo" would be better.
And will there be a link to the PT Settings too, like #7408?

htcfreek commented:
**Additional idea:**
(Follow up my list above.)

10. Multi-mon environment: Maybe it's helpful to have a small symbol overlay on the monitor rectangle to indicate, that a layout is applied on this monitor. @niels9001, I think on using the fz symbol from settings' nav bar.

niels9001 commented:
@htcfreek @Jay-o-Way I have made some changes to the initial proposal based on your (and @crutkas) feedback. See updated images in original post.

htcfreek commented:
@niels9001 
Thanks for respecting our ideas.

**Some thoughts 🤔**
I think about if the new edit symbol shown on hover the layout is a bit to big and hides the view of the layout to much?

How are the keyboard commands for edit and select a layout regarding to accessibility?

I asking me if we need some save and cancel buttons in the new edit layout menu? (Btw. I like the solution we have now.)

**Question ❓**
What do you think about my ideas 6, 7 and 8? (cc: @enricogior)

Jay-o-Way commented:
> I think about if the new edit symbol shown on hover the layout is a bit to big and hides the view of the layout to much?

Yes. Why not use a standard hamburger menu in the corner?

> How are the keyboard commands for edit and select a layout regarding to accessibility?

Maybe a right-click to trigger the context menu? And will it also be touch-screen friendly? (I know a touchscreen is not very common for heavy users, but just curious)

Jay-o-Way commented:
One more thing: in the top of the Editor we read "choose layout for this desktop, but in this area you can choose the **desktop**. The layouts are chosen below that.

niels9001 commented:
@jay-o-way That string will be removed completely. Doesn't add a lot to have the message, the UI is hopefully self-explanatory :).

Jay-o-Way commented:
Okay, two more small things then:
You used "<kbd>+</kbd> (number) <kbd>-</kbd>" for the number of zones in a layout, but in left-to-right writing, the opposites is more natural. And would a [RepeatButton](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.repeatbutton?view=winrt-19041) class be good for these two buttons?

_(apologies if these small but often messages of mine are not easy to work with or something)_

crutkas commented:
Giving this a whirl right now for feedback

crutkas commented:
Things i thought were 🔥 is the canvas grid adjustment highlighting.  That was super hot.

**Bug:**
- Clicked `create new layout` and it looks like the app crashed, dialog didn't pop on top.
![image](https://user-images.githubusercontent.com/1462282/101847485-3af76100-3b08-11eb-8036-5244fdb8fe41.png)

- Grid system, shift doesn't change splitter orientation
![shift not working](https://user-images.githubusercontent.com/1462282/101847733-bfe27a80-3b08-11eb-8974-1b7a3d910576.gif)

- Click Duplicate and dialog stays up
![image](https://user-images.githubusercontent.com/1462282/101847814-e99ba180-3b08-11eb-81c2-3c33cfe67e72.png)

 - changing space doesn't actually seem to reflect change
![image](https://user-images.githubusercontent.com/1462282/101847861-0cc65100-3b09-11eb-86e4-aac049c987fa.png)


**Things to create a new issue on:**
- On mouse off for previewing, it should go back to selected item, see gif
![toggling](https://user-images.githubusercontent.com/1462282/101847293-c0c6dc80-3b07-11eb-997a-f40cc70081c4.gif)

- Small but the Font weight between editor and monitor is diff, i prefer bolder
![image](https://user-images.githubusercontent.com/1462282/101847934-31bac400-3b09-11eb-9507-3902829c3a3c.png)``
![image](https://user-images.githubusercontent.com/1462282/101847988-4e56fc00-3b09-11eb-9800-685003eecc34.png)

**Things for community feedback**
- the monitors on top are far smaller, may be overlooked.  Less in your face than 0.27 i feel like.
- this icon makes me think it can do something.
![image](https://user-images.githubusercontent.com/1462282/101848067-75adc900-3b09-11eb-8500-8a3e1287cc3b.png)

htcfreek commented:
@niels9001 

Thoughts on the actual layout:
![image](https://user-images.githubusercontent.com/1462282/101847485-3af76100-3b08-11eb-8036-5244fdb8fe41.png)
_Image borrowed from @crutkas' comment above._

1. I don't like the new position of the zones count. I personally think it's better on the left side as before or between header and first layout line. On the new position it looks a bit lost.

2. I would optimize the no-custom-layout text a bit: "Create or duplicate a layout to get started **with your own layout**". Reason: You can also start by using one of the temple layouts without creating your own.

enricogior commented:
@crutkas 
> Grid system, shift doesn't change splitter orientation

It has been fixed in master, a rebase should bring the fix here as well.

enricogior commented:
@crutkas 
> changing space doesn't actually seem to reflect change

Code behind for this doesn't exist yet.

niels9001 commented:
> You used "+ (number) -" for the number of zones in a layout, but in left-to-right writing, the opposite (-/+) is more natural. 

Good catch! Fixed.

> * Clicked `create new layout` and it looks like the app crashed, dialog didn't pop on top.

Yeah, it doesn't center correctly if a MaxWidth / MaxHeight is not defined. Filed an issue against it in the ModernWPF repo: https://github.com/Kinnara/ModernWpf/issues/222

> * Small but the Font weight between editor and monitor is diff, i prefer bolder

Fixed.

> * the monitors on top are far smaller, may be overlooked.  Less in your face than 0.27 i feel like.

Yeah, to save space since the visualization itself doesn't add a lot (e.g. could even be a ComboBox). Let's see if it's clear enough or missed by users. The tooltip does provide additional information such as the monitor number and resolution.

> * this icon makes me think it can do something.

Reverted back to the default FZ editor icon.



> @niels9001

> 1. I don't like the new position of the zones count. I personally think it's better on the left side as before or between header and first layout line like the create new link. On the new position (same line as template header) it looks a bit lost.

@htcfreek I see your point. I think the idea is to move these buttons to the flyout, so that they can be set individually per layout. But this won't happen for 0.29, so sticking to the initial position for now.

> 2. I would optimize the no-custom-layout text a bit: "Create or duplicate a layout to get started **with your own layout**". Reason: You can also start by using one of the temple layouts without creating your own.

Yeah, we can update that... but I let @crutkas handle the UI text :).

Jay-o-Way reviewed (commented):
In CanvasZone.xaml, line 125, I see icon `E894` "X" in a button that's marked "Delete Zone". Would `E74D` (trashcan) not be better?
In CanvasEditorWindow.xaml I see icon `E109` "+". On [MS's list](https://docs.microsoft.com/en-us/windows/uwp/design/style/segoe-ui-symbol-font) I read
>  Note: Glyphs with prefixes ranging from E0- to E5- (e.g. E001, E5B1) are currently marked as legacy and are therefore deprecated.

Would that matter? `E710` would be a good substitute.

enricogior commented:
@niels9001 
the layout context menu is it supposed to work? Right click on a template doesn't show it.

niels9001 commented:
> @niels9001
> ~the layout context menu is it supposed to work? Right click on a template doesn't show it.~
> Never mind, I didn't realize that is not right-click but just click on the edit icon.

Yeah, it doesn't.. but I think it would be nice to have the same behavior as clicking the edit button with right-mouse click when the layout is applied. But not a priority I guess.

enricogior commented:
@niels9001 
I think we need to make the zone border more visible, in some cases it's really difficult to see it:

![image](https://user-images.githubusercontent.com/3206696/102364923-4b4f8600-3fb7-11eb-8742-6e272f6f3053.png)

![image](https://user-images.githubusercontent.com/3206696/102365026-6b7f4500-3fb7-11eb-91eb-cf24170a8377.png)

niels9001 commented:
> lor we already use for the zones whe

Good example. This needs to be improved, yeah. Where was this screenshot taken? When hovering over a layout (i.e. LayoutPreview) or in general usage? Dark of light theme?

enricogior commented:
@niels9001 
> Where was this screenshot taken?

Layout preview in the editor.

enricogior commented:
@niels9001 @SeraphimaZ 
I think we need to change the term "Duplicate" to something else, since this action creates a custom layout from the template, it doesn't duplicate the template.

![image](https://user-images.githubusercontent.com/3206696/103203727-dd2f8b00-48f5-11eb-8ff4-246326cbed6f.png)

It should be something like "Create custom", better terms are welcomed.

enricogior commented:
@niels9001 @SeraphimaZ 
in single monitor mode we have a usability issue that needs to be fixed: you can only edit the layout properties if the layout is selected
![image](https://user-images.githubusercontent.com/3206696/103204560-e7eb1f80-48f7-11eb-866d-9b0835ba1545.png)

For all other layouts is not possible to edit the properties without first applying the layout (that closes the editor) then reopening the layout to edit that layout.
My suggestion is that we don't close the editor when applying a layout in single monitor (same as for multi monitor), so it will be possible to edit the properties of a layout before closing the editor.

htcfreek commented:
> @niels9001 @SeraphimaZ 
> for the templates context menu, I think we need to change the term "Duplicate" to something else, since this action creates a custom layout from the template, it doesn't duplicate the template.
> 
> ![image](https://user-images.githubusercontent.com/3206696/103203727-dd2f8b00-48f5-11eb-8ff4-246326cbed6f.png)
> 
> It should be something like "Create custom", better terms are welcomed.
> For custom layouts, the term "Duplicate" is appropriate.

What about the term "Create custom layout" with the tooltip "Create a custom layout based on this template."?

htcfreek commented:
> @niels9001 @SeraphimaZ 
> in single monitor mode we have a usability issue that needs to be fixed: you can only edit the layout properties if the layout is selected
> ![image](https://user-images.githubusercontent.com/3206696/103204560-e7eb1f80-48f7-11eb-866d-9b0835ba1545.png)
> 
> For all other layouts is not possible to edit the properties without first applying the layout (that closes the editor) then reopening the layout to edit that layout.
> My suggestion is that we don't close the editor when applying a layout in single monitor (same as for multi monitor), so it will be possible to edit the properties of a layout before closing the editor.
> 

Can we hide the grey space at the top in single monitor mode? I think it looks like a bug where some ui elements are missing or not shown.

Jay-o-Way commented:
The end of the month is close. Would this PR make it into the next update?

enricogior commented:
This PR depends on https://github.com/microsoft/PowerToys/issues/8779 in order to re-enable the file watcher reload the layout settings.

SeraphimaZ commented:
The work remaining:

P0: The create new layout dialog renders semi-off the screen.
P1: On mouse leave of the layout item, the preview should reset to the applied layout.
P2: Move zones + and - buttons to the template layouts flyout (release vNext++)
P2: Disable layouts and add information tooltip describing that the layout cannot be applied to the current screen resolution.

enricogior commented:
@SeraphimaZ 
regarding:
> P1: On mouse leave of the layout item, the preview should reset to the applied layout.

We should consider disabling the preview in the first place, since it's not really useful and can cause confusion since the current UI doesn't let change any setting for a preview layout.

enricogior commented:
@SeraphimaZ 

>P2: Move zones + and - buttons to the template layouts flyout (release vNext++)
P2: Disable layouts and add information tooltip describing that the layout cannot be applied to the current screen resolution.

Those two should not be considered a blocked to merge this PR.

htcfreek commented:
- Do I have to apply a layout befor I can see the edit symbol? How is it handled at the moment?

- In addition to [my comment above](https://github.com/microsoft/PowerToys/pull/8148#issuecomment-751695254): Why do we hide the monitor preview in single monitor mode? Can we show only the one existing monitor (as selected) in single monitor mode. Then we don't have lost space.

enricogior commented:
@htcfreek 
this is still work in progress and we are considering significant changes.

> Do I have to apply a layout befor I can see the edit symbol? How is it handled at the moment?

This is one of the changes that we already decide to do.

> Why do we hide the monitor preview in single monitor mode?

We will not.

htcfreek commented:
> @htcfreek 
> this is still work in progress and we are considering significant changes.
> 
> > Do I have to apply a layout befor I can see the edit symbol? How is it handled at the moment?
> 
> This is one of the changes that we already decide to do.

👍

> > Why do we hide the monitor preview in single monitor mode?
> 
> We will not.
> 

Maybe we missunderstood each other. But on some images in previous comments there is no monitor symbol in the top space of the window.

niels9001 commented:
@htcfreek as @enricogior pointed out, there will be changes. That's one of the reasons this PR is still in draft.

enricogior commented:
@htcfreek 
> Maybe we missunderstood each other. But on some images in previous comments there is no monitor symbol in the top space of the window.

Because it was a partial screenshot.

Jay-o-Way commented:
> I think we need to change the term "Duplicate" to something else, since this action creates a custom layout from the template, it doesn't duplicate the template.

Strictly speaking you are correct, "duplicate" means "to make an exact copy of something". In practice, it often means "duplicate-and-then-edit".

niels9001 commented:
Closing this in favor of #9325
